File: 백준/Silver/20546. 🐜 기적의 매매법 🐜/🐜 기적의 매매법 🐜.java
import java.io.*;
import java.util.*;

public class Main {
    static int n;
    static int[][] graph; 
    static boolean[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int cash = Integer.parseInt(br.readLine());
        ArrayList<Integer> prices  = new ArrayList<>();
        String[] input = br.readLine().split(" ");
        for (String s : input) {
            prices.add(Integer.parseInt(s));
        }
        int bnp_cash = cash;
        int bnp_stock = 0;
        for(int i = 0;i<prices.size();i++){
            int price = prices.get(i);
            if(bnp_cash>=price){
                int buy = bnp_cash/price;
                bnp_stock+= buy;
                bnp_cash-= buy*price;
            }
        }
        int bnp_total = bnp_cash+bnp_stock*prices.get(prices.size()-1);

        int timing_cash = cash;
        int timing_stock = 0;
        int up_cnt = 0;
        int down_cnt =0;

        for(int i=0;i<prices.size();i++){
            if(i>0){
                if(prices.get(i)>prices.get(i-1)){
                    up_cnt++;
                    down_cnt = 0;
                }else if(prices.get(i)<prices.get(i-1)){
                    down_cnt++;
                    up_cnt = 0;
                }else{
                    up_cnt = 0;
                    down_cnt = 0;
                }
            }
            if(up_cnt>=3 && timing_stock>0){
                timing_cash += timing_stock*prices.get(i);
                timing_stock =0;
            }else if(down_cnt>=3 && timing_cash>=prices.get(i)){
                int buy = timing_cash/prices.get(i);
                timing_stock += buy;
                timing_cash -= buy*prices.get(i);
            }
        }
        int total_timing = timing_cash+timing_stock*prices.get(prices.size()-1);
       
       if(bnp_total>total_timing){
        System.out.println("BNP");
       }else if(bnp_total<total_timing){
        System.out.println("TIMING");
       }else{
        System.out.println("SAMESAME");
       }
    }
}

File: 백준/Silver/20546. 🐜 기적의 매매법 🐜/🐜 기적의 매매법 🐜.py
import sys
input = sys.stdin.readline

cash = int(input().strip())
prices = list(map(int,input().split()))

bnp_cash = cash
bnp_stock = 0
for p in prices:
    if bnp_cash>=p:
        buy = bnp_cash//p
        bnp_stock += buy
        bnp_cash -= buy*p
bnp_total = bnp_cash+bnp_stock *prices[-1]

timing_cash = cash
timing_stock = 0
up_cnt = 0
down_cnt = 0
for i in range(len(prices)):
    if i>0:
        if prices[i]>prices[i-1]:
            up_cnt +=1
            down_cnt = 0
        elif prices[i]<prices[i-1]:
            down_cnt+=1
            up_cnt =0
        else:
            up_cnt = 0
            down_cnt = 0
    if up_cnt>=3 and timing_stock >0:
        timing_cash += timing_stock*prices[i]
        timing_stock = 0
    elif down_cnt>=3 and timing_cash>= prices[i]:
        buy = timing_cash//prices[i]
        timing_stock += buy
        timing_cash -= buy*prices[i]
timing_total = timing_cash + timing_stock*prices[-1]

if bnp_total > timing_total:
    print("BNP")
elif bnp_total < timing_total:
    print("TIMING")
else:
    print("SAMESAME")